skxray/correlation.py

```diff
@@ -0,0 +1,375 @@
+                                       level=level, buf_no=cur[level]-1,)
+                level += 1
+
+                # Checking whether there is next level for processing
+                if level<num_levels:
+                    processing = True
+                else:
+                    processing = False
+            else:
+                cts[level] = 1
+                processing = False
+
+    # ending time for the process
+    t2 = time.time()
+
+    logger.info("Processing time for {0} images took {1} seconds."
+                "".format(len(img_stack), (t2-t1)))
+
+    # to get the final G, IAP and IAF values
+    g_max = IAP.shape[0]
+
+    # calculate the one time correlation
+    g2 = (G[: g_max] / (IAP[: g_max] * IAF[: g_max]))
+
+    # finding the lag times (delay times) for multi-tau levels
+    tot_channels, lag_steps = core.multi_tau_lags(num_levels,
+                                                  num_bufs)
+    lag_steps = lag_steps[:g2.shape[0]]
+
+    return g2, lag_steps
+
+
+def _process(buf, G, IAP, IAF, q_inds, num_bufs,
+             num_pixels, num, level, buf_no):
+    """
+    This helper function calculates G, IAP and IAF at
+    each level, symmetric normalization is used.
+
+    Parameters
+    ----------
+    buf : ndarray
+        image data array to use for correlation
+
+    G : ndarray
+        matrix of auto-correlation function without
+        normalizations
+
+    IAP : ndarray
+        matrix of past intensity normalizations
+
+    IAF : ndarray
+        matrix of future intensity normalizations
+
+    q_inds : ndarray
+        indices of the required roi's
+
+    num_bufs : int, even
+        number of buffers(channels)
+
+    num_pixels : ndarray
+        number of pixels in certain roi's
+        roi's, dimensions are : [num_qs]X1
+
+    num : ndarray
+        to track how many images processed in each level
+
+    level : int
+        the current level number
+
+    buf_no : int
+        the current buffer number
+
+    Returns
+    -------
+    G : ndarray
+        matrix of auto-correlation function without normalizations
+
+    IAP : ndarray
+        matrix of past intensity normalizations
+
+    IAF : ndarray
+        matrix of future intensity normalizations
+
+    Notes
+    -----
+    :math ::
+        G   = <I(t)I(t + delay)>
+
+    :math ::
+        IAP = <I(t)>
+
+    :math ::
+        IAF = <I(t + delay)>
+
+    """
+    num[level] += 1
+
+    # in multi-tau correlation other than first level all other levels
+    #  have to do the half of the correlation
+    if level == 0:
+        i_min = 0
+    else:
+        i_min = num_bufs//2
+
+    for i in range(i_min, min(num[level], num_bufs)):
+        t_index = level*num_bufs/2 + i
+
+        delay_no = (buf_no - i) % num_bufs
+
+        IP = buf[level, delay_no]
+        IF = buf[level, buf_no]
+
+        G[t_index] += ((np.bincount(q_inds,
+                                    weights=np.ravel(IP*IF))[1:])/num_pixels
```

Comment:
Ah, I was misreading the `()` and thought you were slicing on `(IP*IF)[1:]`  I would break this up into two steps

``python
tmp_binned = (np.bin_count(...)[1:]) / num_pixels
G[t_index] += tmp_binned
```